### PR TITLE
add some more stats

### PR DESCRIPTION
**Pull request by russelltg**
Adds: 
- tx_buffered_time
- tx_buffered_data
- tx_buffered_bytes
- rx_acknowledged_time
- tx_all_bytes
- tx_light_ack
- rx_retransmit_data
- rx_unique_data
- rx_unique_bytes
- rx_light_ack
- tx_dropped_data
- tx_unique_data
- tx_retransmit_data

### CODE REVIEW

**Comment by russelltg** [comment]
@robertream I was writing a test for my sender buffer metrics and sender/buffer.rs:43:

```rust
        if self.buffer.is_empty() {
            self.buffer.push_back(packet.clone());
        }
```

What's up with this? What purpose does this serve? If it's necessary it would be good to have a comment here, but ideally this isn't necessary?

**Comment by robertream** [comment]
> @robertream I was writing a test for my sender buffer metrics and sender/buffer.rs:43:
> 
> ```rust
>         if self.buffer.is_empty() {
>             self.buffer.push_back(packet.clone());
>         }
> ```
> 
> What's up with this? What purpose does this serve? If it's necessary it would be good to have a comment here, but ideally this isn't necessary?

Yes, agreed, it deserves a comment, though this is for the edge case of stream ending, where one remaining packet needs to be sent for the integration tests to pass. There are several places where this hack is implemented, some of them commented, some of them not. I believe there is also a unit test covering it. Should we no longer need the work around, we'd also need to clean up the entire work around, not just these lines.

**Comment by russelltg** [comment]
Will the hack no longer be necessary after we implement RTO?

**Comment by robertream** [comment]
> Will the hack no longer be necessary after we implement RTO? Or am I not understanding properly

I recall trying to fix this in another way, but it was never clear what the underlying cause of the problem was. The one failing test that I recall working on was expecting just one packet sent and received, so I assume it is some kind of race condition, where the sender socket is probably closing before the packet is ACKed.

**Comment by robertream** [comment]
> > Will the hack no longer be necessary after we implement RTO? Or am I not understanding properly
> 
> I recall trying to fix this in another way, but it was never clear what the underlying cause of the problem was. The one failing test that I recall working on was expecting just one packet sent and received, so I assume it is some kind of race condition, where the sender socket is probably closing before the packet is ACKed.

So it is unclear to me if RTO will fix it, but could, maybe, depending on how we implement it.

**Comment by russelltg** [comment]
Gotcha, well I'll leave it for now. I think this PR is in a decent state, would be great to get a review

**Comment by russelltg** [comment]
@robertream Could you review this? it's pretty small, but making sure I'm doing this more or less the way you had in mind

File: srt-protocol/src/protocol/receiver/buffer.rs [hunk 1]
```diff
@@ -503,6 +503,21 @@ impl ReceiveBuffer {
...
+        if let (Some(BufferPacket::Received(s)), Some(BufferPacket::Received(e))) =
+            (self.buffer.get(start_idx), self.buffer.get(end_idx))
+        {
+            Duration::from_micros(
+                u64::try_from((e.timestamp - s.timestamp).as_micros()).unwrap_or(0),
+            )
+        } else {
+            Duration::from_micros(0)
+        }
+    }
```
Comment (robertream) [inline srt-protocol/src/protocol/receiver/buffer.rs:520]: is there a unit test that covers this?

**Review by robertream — APPROVED** [review]
looks good to me

File: srt-protocol/src/packet/control/srt.rs [hunk 2]
```diff
@@ -304,6 +304,14 @@ fn string_to_le_bytes(str: &str, into: &mut impl BufMut) {
     }
 }
 
+fn generate_filter_string(filter: &BTreeMap<String, String>) -> String {
```
[the remark this answers is not in the document]
Comment (russelltg) [reply srt-protocol/src/packet/control/srt.rs]: Agreed, I have a step in this direction by having a FilterSpec type, and making that a Display implementation
